Makefile.cflags: Build with -Wall -Werror by default (including fixes to correct all existing warnings)

gebart opened the pull request:
Fixes a lot of warnings. This is a stepping stone towards #1121.

`make buildtest` with `CFLAGS += -Wall -Werror` (not `-Wextra`) added to `Makefile.include` now completes without errors on ~~`examples/hello-world`, `examples/default`, `examples/rpl_udp`, `tests/unittests`~~ all tests

Worth extra notice:
 - openmote GPIO periph_conf has been redefined to allow all pins as GPIOs (before only the few first were defined)
 - redbee-econotag does not set `CFLAGS += -Wextra` in the board Makefile anymore
 - msp430 now has a wrapper stdlib.h header for including malloc et al. as expected from `<stdlib.h>`

Kijewski commented on cpu/cortexm_common/cortexm_init.c:
```diff
@@ -46,7 +46,7 @@ void cortexm_init(void)
     /* initialize all vendor specific interrupts with the same value */
-    for (int i = 0; i < CPU_IRQ_NUMOF; i++) {
-        NVIC_SetPriority(i, CPU_DEFAULT_IRQ_PRIO);
+    for (int i = 0; i < (int) CPU_IRQ_NUMOF; i++) {
```
`s/int i/IRQn_Type i/`?

gebart replied:
addressed

gebart commented:
Added -Wall -Werror as an experiment, waiting for Travis.

gebart commented:
Added a whole bunch of fixes.

Interesting parts:
 - msp430 getchar implementations for all msp430 boards except chronos
 - Blacklist MSP430 boards for libfixmath_unittests because of some missing math.h functions.
 - Blacklist MSP430 from CCN lite examples because the API seem to be designed with the assumption that `int` is larger than 16 bits wide.
 - Added libfixmath patches for fixing warnings inside the library.

gebart commented:
there's a problem with the qemu-i386 port and PRId64 being defined as `d` instead of `lld` because of a missing declaration of `__have_long_long_64`. I am guessing that a toolchain update might fix it but I am unfamiliar with the toolchain script.

OlegHahm commented:
Awesome work! Thanks a lot!

>  I am guessing that a toolchain update might fix it but I am unfamiliar with the toolchain script.

Which toolchain script are you referring to?

gebart commented:
@OlegHahm I was referring to the script in `dist/tools/toolchain/build_x86.sh`.

`boards/x86-multiboot-common/Makefile.include` downloads a precompiled and patched newlib from http://download.riot-os.org/i586-newlib_2.1.0_tlsf-1254.tar.bz2. I am assuming this newlib was built via the build_x86.sh script.
The build_x86.sh script in turn has a bunch of cryptic patches without explanation what they are for or what they do, see https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/toolchains/build_x86.sh#L30

I think the problem would be solved by an upgrade to newlib-2.2.0, which has some updates for the PRI* macros.

OlegHahm commented:
@Kijewski, can you shed some light on this?

Kijewski commented:
Yes, the changes are: 
 - supplying my own malloc
 - un-inlining (f)putc
 - fixing some minor warnings

gebart commented:
nevermind my comment about newlib versions.
The real problem is that `boards/x86-multiboot-common/Makefile.include` is adding `-isystem /usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/include` before the newlib includes, which means that my x86_64 libgcc stdint.h is picked instead of newlib's stdint.h. GCC does not supply the macros `__have_long64` etc.

When supplying `-ffreestanding` as an argument to GCC, the compiler will define `__STDC_HOSTED__` as 0 and include its own headers instead of the libc's (newlib's)

What is the correct procedure here? Should newlib's includes be added before GCC's on the command line? Is it possible to build a QEMU kernel without `-ffreestanding`?

gebart commented:
Right after I posted my last comment I found this mailing list post:

http://comments.gmane.org/gmane.comp.lib.newlib/9373

(The change discussed in the post is included in newlib 2.2.0, but not in newlib 2.1.0)

gebart commented:
Might need an even newer newlib:
https://sourceware.org/ml/newlib/2015/msg00406.html

gebart commented:
Rebased on latest master.

 - x86: Rename `stack_t` -> `stack_info_t` because recent newlib defines `stack_t` inside `<sys/signal.h>`.

See https://sourceware.org/ml/newlib-cvs/2015-q2/msg00001.html

The qemu-i386 tests work now if I replace the files in `toolchain/x86/i586-none-eabi/include` by the include files from my installed newlib (from git some weeks old). All PRId64, PRIu32 bugs have been fixed in newlib now.

OlegHahm commented on sys/net/include/etx_beaconing.h:
```diff
@@ -51,13 +50,13 @@ extern "C" {
  * and 5 (For ETX_MAX_JITTER) unless those values are adjusted too.
  */
-#define ETX_INTERVAL        (1000)
+#define ETX_INTERVAL        (1000ul)
```
Why not replacing with `MS_IN_USEC` while you're at it anyway?

OlegHahm commented on tests/bitarithm_timings/main.c:
```diff
@@ -34,8 +34,8 @@
-#define TIMEOUT_S (5)
-#define TIMEOUT_US (TIMEOUT_S * 1000 * 1000)
+#define TIMEOUT_S (5ul)
+#define TIMEOUT_US (TIMEOUT_S * 1000ul * 1000ul)
```
`SEC_IN_USEC` instead?

gebart replied:
good idea

OlegHahm commented:
As a "side effect" this PR fixes the UART on MSP430 platforms - at least on the Z1 I'm currently using.

OlegHahm commented:
Apart from the minor comments: ACK

gebart commented:
Don't merge until we have updated newlib on Travis for qemu-i386

gebart commented:
opened #3337 to upgrade the x86 newlib binary tarball

gebart commented:
@OlegHahm comments addressed

gebart commented:
OK to squash the comment commits?

OlegHahm commented on tests/bitarithm_timings/Makefile:
```diff
@@ -1,4 +1,7 @@
 APPLICATION = bitarithm_timings
+
+USEMODULE+=timex
```
This shouldn't be necessary just for the include.

gebart replied:
addressed, OK to squash immediately?

OlegHahm replied:
Yes, please, go ahead!

kaspar030 replied:
> OK to squash immediately?

yes!

gebart commented:
addressed comment by @OlegHahm (removed USEMODULE timex in bitarithm_timings)

gebart commented:
squashed

gebart commented:
Added a temporary dropbox link for an i586-elf compiled newlib 2.2.0.20150623.

Kijewski commented on Makefile.cflags:
```diff
@@ -51,6 +51,9 @@ endif
 CFLAGS += -fno-common
 
+# Enable all default warnings
+CFLAGS += -Wall
```
Actually I am not really sure if this is the right place to add this argument. It prevents the user from writing crappy code, which they might not take kindly.

gebart replied:
Enabling the warnings does not turn them into errors. `-Werror` is only enabled in `make buildtest`

Kijewski replied:
Ahh, right. Sorry. :)

Kijewski commented on sys/posix/pthread/pthread_spin.c:
```diff
@@ -47,7 +47,7 @@ int pthread_spin_lock(pthread_spinlock_t *lock)
     }
 
-    while (atomic_set_to_one((int *)lock) == 0) {
+    while (atomic_set_to_one((atomic_int_t *)lock) == 0) {
```
Somehow I don't like this typecast but I can't really say why. Maybe change the def to:
```c
typedef struct {
    atomic_int_t value;
} pthread_spinlock_t;
```
?

gebart replied:
Sure, I'll update it.

gebart commented:
Rebased.
How do we go about merging this? 
Should I squash anything other than the commits with message "`squash`" @OlegHahm @Kijewski ?

OK to squash?

gebart commented:
Rebased after updating the x86 toolchain.

OlegHahm commented:
Still want to squash anything?

gebart commented:
There are a few squash commits in the middle, OK to squash?

OlegHahm commented:
Go ahead!

gebart commented:
How far are we from an ACK after squashing?

gebart commented:
squashed

OlegHahm commented:
My previous ACK still holds.

authmillenon commented:
whitespace errors are introduced to (or revealed in this PR) `pkg/libfixmath/libfixmath-unittests-printf-format.patch`

gebart commented:
@authmillenon it is not possible to avoid since the patch will not apply without the whitespace errors.

gebart commented:
(the problem with trying to maintain another project's code with other coding conventions)

Kijewski commented:
What's the problem with adhering to two conflicting coding conventions? ;-)

authmillenon commented:
It is: use `--ignore-whitespaces` for `git am` ;-)

authmillenon commented:
Same goes for `patch`.

authmillenon commented:
Re-ACK, please squash.

gebart commented:
squashed, waiting for Travis

OlegHahm commented:
Woohoo!

thomaseichinger commented:
This broke building for OS X and FreeBSD. Again :cry:

gebart commented:
>This broke building for OS X and FreeBSD. Again :cry:

Sorry about that!

gebart commented:
@thomaseichinger is FreeBSD and OSX sufficiently similar to assume that any build fixes for FreeBSD usually gets it working on OSX as well?

thomaseichinger commented:
No, sadly not.

authmillenon commented:
It's about time, that Millhouse (Mac@FU) and Stedten (FreeBSD@FU) are included in the build tests :D

thomaseichinger commented:
I fear we wont fix this for OS X, as the used `ucontext` API used by native is marked as deprecated.

gebart commented:
@thomaseichinger we could add an `ifeq` check for OSX in the Makefile and defer fixing this until the next OSX release where the deprecated stuff is removed.

thomaseichinger commented:
I'm on it.

gebart commented:
@thomaseichinger does `-Wno-error=deprecated` get the correct result?

thomaseichinger commented:
@gebart I went with `-Wno-deprecated-declarations`.
